blog/features/replica-signed-queries.md

```diff
@@ -0,0 +1,149 @@
+tags: [New features]
+image: /img/blog/replica-signed-queries.jpg
+---
+
+When an update call is made to a canister on the Internet Computer, the response to includes a subnet signature that proves the data sent within the request is trustworthy. Query calls, however, are handled by single replicas and don't bear any information that validates their response is trustworthy, since their data could be tampered with by a malicious replica or malicious boundary node. As a workaround, certified data and certified variables can be used, which are embedded in responses to query calls that provide a layer of validation. However, using certified data and variables requires extra work for developers and sometimes does not fit into a developer's workflow seamlessly. 
+
+To enhance the trustworthiness of queries by providing better guarantees against malicious attacks, we're excited to announce the release of replica signed queries! 
+
+## What are replica signed queries?
+
+Replica signed queries are a new feature that introduces replica signatures in query responses. Every response to query call will contain a signature that signs specific parts of the response using the replica's private key. Then, query responses can be validated by fetching the replica's public key from the subnet's state tree, reading the public key of the signing replica, then verifying the signature.
+
+This new feature includes two primary changes: first, the subnet state tree now exposes the public keys of the replicas, then query responses will now contain a signature. 
+
+Having a single replica signature in a query response doesn't completely solve the problem that query responses cannot be trusted, however, replica signed queries do help improve the security of queries since the recipient can validate that the response came from a replica on the correct subnet. The primary goal of this feature is to provide better guarantees against malicious boundary nodes, since replica signed queries make it impossible for a malicious boundary node to return false data. 
+
+## How do replica signed queries work?
+
+First, a list of node IDs and their corresponding public keys is put into the subnet's state tree. The `node_signing_public_key` value is used for signing query replies. This value is stored in the registry, then populated into the ReplicatedState field `NodePublicKeys` through a `get_crypto_key_for_node` call to the registry that returns the public keys for the subnet's nodes. 
+
+Users can obtain the certificate to validate the node's keys through an HTTP `read_state` call that includes a timestamp value. 
+
+To support replica signed queries, the query response format has been changed. Previously, query responses used the following format:
+
+```
+pub enum HttpQueryResponse {
+    Replied {
+        reply: HttpQueryResponseReply,
+    },
+    Rejected {
+        error_code: String,
+        reject_code: u64,
+        reject_message: String,
+    },
+}
+```
```

Comment:
I generally suggest removing all the code from the post for several reasons. 
First, they are for internal implementation in certain languages. 
Second, these code are drafted in the very early stage of feature designing and implementation may vary. 
Last, we have formal and canonical representation of all the data types in our interface specification.  https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query

If we want to keep this, it is better to quote the corresponding part in the specification. I will add links for each of them now.

Reply:
```
If the query call resulted in a reply, the response is a CBOR (see [CBOR](https://internetcomputer.org/docs/current/references/ic-interface-spec/#cbor)) map with the following fields:

status (text): "replied"

reply: a CBOR map with the field arg (blob) which contains the reply data.

signatures ([+ node-signature]): a list containing one node signature for the returned query response.

If the call resulted in a reject, the response is a CBOR map with the following fields:

status (text): "rejected"

reject_code (nat): The reject code (see [Reject codes](https://internetcomputer.org/docs/current/references/ic-interface-spec/#reject-codes)).

reject_message (text): a textual diagnostic message.

error_code (text): an optional implementation-specific textual error code (see [Error codes](https://internetcomputer.org/docs/current/references/ic-interface-spec/#error-codes)).

signatures ([+ node-signature]): a list containing one node signature for the returned query response.
```